### PR TITLE
Ys signed vint

### PR DESCRIPTION
**Pull request by johnjohnlin**
vint supports both signed and unsigned now.

### CODE REVIEW

File: common/include/verilog_int.h [hunk 1]
```diff
@@ -450,38 +466,82 @@ struct vint {
...
+
+	// >>=, signed
+	vint& sshiftopr(const unsigned rhs) {
+		const bool is_neg = Bit(num_bit-1u);
+		raw_shiftopr(rhs, is_neg);
+		return *this;
+	}
+
+	vint& operator>>=(const unsigned rhs) {
+		if constexpr (is_signed) {
```
Comment (yodalee) [inline common/include/verilog_int.h:512]: How about just a one `shiftopr`, with argument is_signed.
Then `is_neg = is_signed && Bit(num_bit-1)`

File: common/include/verilog_int.h [hunk 2]
```diff
@@ -450,38 +466,82 @@ struct vint {
...
+			sshiftopr(rhs);
+		} else {
+			ushiftopr(rhs);
+		}
 		return *this;
 	}
 
+	[[gnu::noinline]]
```
Comment (yodalee) [inline common/include/verilog_int.h:520]: Why do we need noinline after this change?

**Review by yodalee — COMMENTED** [review]
It is quite complicated, I think it is fine if all the tests are passed.
I have some suggested minor fixes, you can fix them and merge the changes by squash it.

File: test/test_verilog_int.cpp [hunk 3]
```diff
@@ -106,31 +107,68 @@ void RuleOfThreeTemplate() {
...
+		// Forcely no sign extension
+		vsint<99> v99(-123, false);
+		EXPECT_EQ(v99.v[0], uint64_t(-123));
+		EXPECT_EQ(v99.v[1], uint64_t(0));
+	}
+	vsint<99> v99(0);
+	// No sign extension by default
+	v99 = -1;
+	EXPECT_EQ(v99.v[0], uint64_t(-1));
+	EXPECT_EQ(v99.v[1], uint64_t(0x7ffffffffllu));
```
[the remark this answers is not in the document]
Comment (johnjohnlin) [reply test/test_verilog_int.cpp:176]: This is fixed in a newer commit.

File: common/include/verilog_int.h (hunk 2 above)
Comment (johnjohnlin) [reply common/include/verilog_int.h:520]: I think scattering large inline functions everywhere will harm the performance (long compile time, large generated executable, Icache miss). However, I am not sure whether it is a plus; need more profiling results. How about just let it remain here for the time being.

File: common/include/verilog_int.h (hunk 1 above)
Comment (johnjohnlin) [reply common/include/verilog_int.h:512]: I think this is good. Let me unify the shiftopr/assign/uvalue

File: common/include/verilog_int.h (hunk 2 above)
Comment (yodalee) [reply common/include/verilog_int.h:520]: Sure.

**Pull request merged**